substrate/client/network/sync/src/strategy/state.rs

```diff
@@ -304,10 +306,21 @@ impl<B: BlockT> StateStrategy<B> {
 		let threshold = std::cmp::max(median, min_best_number);
 		// Find a random peer that is synced as much as peer majority and is above
 		// `min_best_number`.
-		for (peer_id, peer) in self.peers.iter_mut() {
-			if peer.state.is_available() && peer.best_number >= threshold {
-				peer.state = new_state;
-				return Some(*peer_id)
+		for mut available_peer in self.peer_pool.lock().available_peers() {
+			let peer_id = available_peer.peer_id();
+			if let Some(peer) = self.peers.get_mut(peer_id) {
+				if peer.best_number >= threshold {
+					available_peer.reserve();
+					peer.state = new_state;
+					return Some(*peer_id)
+				}
+			} else {
+				warn!(
+					target: LOG_TARGET,
+					"State inconsistency: peer {peer_id} is in the pool of connected peers, \
+					 but not known to `StateStrategy`.",
+				);
+				debug_assert!(false);
```

Comment:
What about flipping testing the other way: this code iterates over its peers, check if a peer satisfies `peer.best_number >= threshold` and then it calls `PeerPool::try_reserve_peer()` and if returns true, the peer is reserved and if returns false, it continues to the next peer. I think it would clean up this code and would hold the lock only while trying to reserve a peer.

Reply:
We can do that, but now it doesn't make a big difference as requests are generated after `select!` in `SyncingEngine` where all the strategies are called serially, so the lock is always available.

Reply:
Would get rid of this `else` branch with `debug_assert!(false)` though

Reply:
Done